src/preprocessing/passes/bool_to_bv.cpp

```diff
@@ -39,165 +39,159 @@ PreprocessingPassResult BoolToBV::applyInternal(
+  {
+    d_lowerCache[n] = rebuilt;
+  }
+  else
+  {
+    d_iteBVLowerCache[n] = rebuilt;
+  }
+}
+
 Node BoolToBV::fromCache(TNode n) const
 {
-  if (d_lowerCache.find(n) != d_lowerCache.end())
+  // check more likely case first
+  if (n.getKind() != kind::ITE)
+  {
+    if (d_lowerCache.find(n) != d_lowerCache.end())
+    {
+      return d_lowerCache.at(n);
+    }
+  }
+  else
   {
-    return d_lowerCache.find(n)->second;
+    if (d_iteBVLowerCache.find(n) != d_iteBVLowerCache.end())
+    {
+      return d_iteBVLowerCache.at(n);
+    }
   }
   return n;
 }
 
+inline bool BoolToBV::inCache(const Node& n) const
+{
+  return (ContainsKey(d_lowerCache, n) || ContainsKey(d_iteBVLowerCache, n));
+}
+
 bool BoolToBV::needToRebuild(TNode n) const
 {
   // check if any children were rebuilt
   for (const Node& nn : n)
   {
-    if (ContainsKey(d_lowerCache, nn))
+    if (inCache(nn))
     {
       return true;
     }
   }
   return false;
 }
 
-Node BoolToBV::lowerAssertion(const TNode& a)
+Node BoolToBV::lowerAssertion(const TNode& assertion, bool force)
+{
+  // first try to lower all the children
+  for (const Node& c : assertion)
+  {
+    lowerNode(c, force);
+  }
+
+  // now try lowering the assertion, but don't force it (even in mode all)
+  lowerNode(assertion, false);
+  Node newAssertion = fromCache(assertion);
+  TypeNode newAssertionType = newAssertion.getType();
+  if (newAssertionType.isBitVector())
+  {
+    Assert(newAssertionType.getBitVectorSize() == 1);
+    NodeManager* nm = NodeManager::currentNM();
+    newAssertion = nm->mkNode(kind::EQUAL, newAssertion, bv::utils::mkOne(1));
+    newAssertionType = newAssertion.getType();
+  }
+  Assert(newAssertionType.isBoolean());
+  return newAssertion;
+}
+
+Node BoolToBV::lowerNode(const TNode& node, bool force)
 {
-  bool optionITE = options::boolToBitvector() == BOOL_TO_BV_ITE;
-  NodeManager* nm = NodeManager::currentNM();
   std::vector<TNode> visit;
-  visit.push_back(a);
+  visit.push_back(node);
   std::unordered_set<TNode, TNodeHashFunction> visited;
-  // for ite mode, keeps track of whether you're in an ite condition
-  // for all mode, unused
-  std::unordered_set<TNode, TNodeHashFunction> ite_cond;
 
   while (!visit.empty())
   {
     TNode n = visit.back();
     visit.pop_back();
 
-    int numChildren = n.getNumChildren();
-    Kind k = n.getKind();
-    Debug("bool-to-bv") << "BoolToBV::lowerAssertion Post-traversal with " << n
-                        << " and visited = " << ContainsKey(visited, n)
+    Debug("bool-to-bv") << "BoolToBV::lowerNode: Post-order traversal with "
+                        << n << " and visited = " << ContainsKey(visited, n)
                         << std::endl;
 
+    int numChildren = n.getNumChildren();
+
     // Mark as visited
-    /* Optimization: if it's a leaf, don't need to wait to do the work */
-    if (!ContainsKey(visited, n) && (numChildren > 0))
+    if (!ContainsKey(visited, n))
```

Comment:
Since this, `if` has an `else`, why not switch the bodies and get rid of the `!`, eliminating the implicit double-negation?